integr8ly/integreatly-operator
# INTLY-8450

# Description
Addressing Jira: https://issues.redhat.com/browse/INTLY-8450 

## What has been done so far
1) New fields in RHMIConfig CRD: 
```
UpgradeAvailable:
availableAt: creationTimestamp
targetVersion: latest avaialable version of the RHMI Operator
```
2) Update status functionality was moved from the subscription controller into RHMIConfig controller
3) Update status function was edited so that it no longer requires installPlan - we used it only to get the creationTime stamp which is now set by the subscription controller and is available in the RHMIConfig.
4) Changed slightly the logic around the `if serviceAffecting` in the subscription controller `handleUpgrades` function
5) Functionality was tested ( although two more steps are left before this PR can be considered as lgtm)
6) RHMIConfig now reflects changes to the user set dates immediately and doesn't need to wait for the subscription reconciler to kick in.

## What needs to be done
1) As the `updateStatus` function has been changed, unit testing needs to be edited as well
2) Seems that I had issues with simulating a patch upgrade, it looks like the current install plan from version N1 to N2 isn't deleted when a catalogSource is pointing at version N1.1 replacing version N1 - There is either an issue with the logic around patching an upgrade ( deleting install plan and recreating it in case of patch upgrade) or I was doing something wrong. 

## To Verify
### Short:
- install RHMI Operator 2.5.0 via catalog source
- simulate an upgrade of the operator to version 2.6.0 
- change `waitForMaintenance= false` and `notBeforeDays=2` in RHMIConfig yaml file
- `upgrade` `scheduled` date should be updated immediately
- change `notBeforeDays=0` 
- upgarde of RHMI Operator should trigger 
- once upgrade is completed, RHMIConfig `upgradeAvailable` and `upgrade` should disappear. 

### Long
1) in integreatly operator repo run:
```
SEMVER=2.5.0 ./scripts/prepare-release.sh
```
2) Make sure to set your own repo name in Makefile https://github.com/integr8ly/integreatly-operator/blob/master/Makefile#L3
3) Run `make image/build/push`
4) Navigate to `/deploy/olm-catalog/integreatly-operator/2.5.0`
5) In `integreatlyoperator.clusterserviceversion` delete the `replaces` field and set the https://github.com/integr8ly/integreatly-operator/blob/master/deploy/olm-catalog/integreatly-operator/2.5.0/integreatly-operator.v2.5.0.clusterserviceversion.yaml#L44  and the https://github.com/integr8ly/integreatly-operator/blob/master/deploy/olm-catalog/integreatly-operator/2.5.0/integreatly-operator.v2.5.0.clusterserviceversion.yaml#L314 to your own repo ie: `quay.io/<your repo name>/integreatly-operator:v2.5.0`
6) Run the following commands: 
```
opm alpha bundle generate -d . --channels alpha --package integreatly --output-dir bundle --default alpha
```
```
docker build -f bundle.Dockerfile -t quay.io/<quay username>/integreatly-bundle:2.5.0 . 
```
```
docker push quay.io/<quay username>/integreatly-bundle:2.5.0
```
```
operator-sdk bundle validate quay.io/<quay username>/integreatly-bundle:2.5.0
```
```
opm index add \
    --bundles quay.io/<your org>/integreatly-bundle:2.5.0 \
    --build-tool docker \
    --tag quay.io/<your org>/integreatly-index:2.5.0
```
```
docker push quay.io/<your org>/integreatly-index:2.5.0
```
7) Make sure you are loggind in to correct cluster 
8) Create file `rhmi-operators.yaml` and paste the following contents:
```
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: rhmi-operators
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: quay.io/<your org>/integreatly-index:2.5.0
```
9) Run `oc apply -f rhmi-operators.yaml`
10) Run 
```
make cluster/prepare/project && \
   make cluster/prepare/smtp && \
   make cluster/prepare/dms && \
   make cluster/prepare/pagerduty
```
11) Navigate to your cluster and in to the Operators Hub, under `openshift-marketplace` you should see RHMI Operator, install it under `rhmi-operator` namespace.
12) Back in your main operator repo run:
```
SEMVER=2.6.0 ./scripts/prepare-release.sh
```
13) Repeat step 2 and 3 but this time navigate to 2.6.0
14) Edit the `replaces` field https://github.com/integr8ly/integreatly-operator/blob/master/deploy/olm-catalog/integreatly-operator/2.5.0/integreatly-operator.v2.5.0.clusterserviceversion.yaml#L459 to `2.5.0` and both images to your own repo as in step 5.
15) Run the following commands:
```
opm alpha bundle generate -d . --channels alpha --package integreatly --output-dir bundle --default alpha
```
```
docker build -f bundle.Dockerfile -t quay.io/<quay username>/integreatly-bundle:2.6.0 . 
```
```
docker push quay.io/<quay username>/integreatly-bundle:2.6.0
```
```
operator-sdk bundle validate quay.io/<quay username>/integreatly-bundle:2.6.0
```
```
opm index add \
    --bundles quay.io/<your org>/integreatly-bundle:2.5.0 \
    --bundles quay.io/<your org>/integreatly-bundle:2.6.0 \
    --build-tool docker \
    --tag quay.io/<your org>/integreatly-index:2.6.0
```
```
docker push quay.io/<your org>/integreatly-index:2.6.0
```
16) Navigate to catalogSource CR under `openshift-marketplace` namespace and change the `  image: quay.io/<your org>/integreatly-index:2.5.0` to `  image: quay.io/<your org>/integreatly-index:2.6.0`
17) Navigate to installed operators and you should see upgrade being triggered soon.
18) Navigate to RHMIConfig CR in RHMI Operator
19) Change the `spec` `notBeforeDays:2` and `waitForMaintenance:false` - changes to RHMIConfig `upgrade scheduled` should happen immediately.
20) Change `notBeforeDays:0` and installation of new version of RHMI Operator should trigger.



## Type of change




## Checklist
